src/cljs-lib/src/calva/repl/webview/core.cljs

```diff
@@ -0,0 +1,229 @@
+   on the given color theme kind."
+  [{:keys [vscode/vscode]} {:keys [color-theme-kind webview-panel]}]
+  (let [color-theme-kind-enum (.. ^js vscode -ColorThemeKind)
+        code-theme (condp = color-theme-kind
+                     (.. color-theme-kind-enum -Dark)  "dark"
+                     (.. color-theme-kind-enum -Light) "light"
+                     (.. color-theme-kind-enum -HighContrast) "high-contrast"
+                     (.. color-theme-kind-enum -HighContrastLight) "high-contrast-light"
+                     nil)]
+    (if code-theme
+      (post-message-to-webview webview-panel {:command/name "set-code-theme"
+                                              :content code-theme})
+      (util/log-to-console
+       :error
+       "Cannot set code theme in output webview. There is no code theme set for the ColorThemeKind enum value of"
+       color-theme-kind))))
+
+(defn create-color-theme-change-listener
+  [{:keys [^js vscode/vscode] :as context}
+   {:keys [webview-panel]}]
+  (.. vscode -window
+      (onDidChangeActiveColorTheme
+       (fn [e]
+         (set-code-theme! context {:color-theme-kind (.. e -kind)
+                                   :webview-panel webview-panel})))))
+
+(defn add-subscriptions!
+  [{vscode-context :vscode/context
+    :as context}
+   {:keys [webview-panel]}]
+  (let [subscriptions [(create-color-theme-change-listener context {:webview-panel webview-panel})]]
+    (run! (fn [subscription]
+            (.. ^js vscode-context -subscriptions (push subscription)))
+          subscriptions)))
+
+(defn create-repl-output-webview-panel
+  [{:keys [vscode/vscode] :as context}]
+  (let [webview-panel (.. ^js vscode -window
+                          (createWebviewPanel
+                           "calva:repl-output"
+                           "REPL Output"
+                           #js {:preserveFocus true
+                                :viewColumn (.. ^js vscode -ViewColumn -Beside)}
+                           #js {:enableScripts true
+                                ;; If performance or memory consumption becomes a problem, we can use the setState
+                                ;; and getState to manually retain the context of the webview when it's hidden.
+                                ;; See https://code.visualstudio.com/api/extension-guides/webview#persistence
+                                :retainContextWhenHidden true
+                                :enableFindWidget true}))]
+    (.. ^js webview-panel (onDidDispose (fn [] (dispose-repl-output-webview-panel repl-output-webview-panel))))
+    (set-webview-html! context {:webview-panel webview-panel})
+    (add-subscriptions! context {:webview-panel webview-panel})
+    webview-panel))
+
+(defn ^:export show-repl-output-webview-panel
+  [preserve-focus?]
+  (let [context {:env/is-debug (:is-debug util/env)
+                 :vscode/vscode @util/vscode
+                 :vscode/context @util/vscode-context}
```

Comment:
The pattern I followed is constructing a context in top-level functions using dereferenced values of the atoms storing the vscode API and the vscode context. This makes testing a little simpler because instead of redef'ing or resetting atoms, we just construct a context with whatever mocks we want. This also makes downstream functions more pure.

Reply:
Also, being able to mock the vscode API and test that certain functions in it are being called correctly it pretty neat 😃.

Reply:
In my experience VS Code complains if I try to deref the API object. You are not running in to that? If so I should have a look at what I do differently...

Reply:
I have not experienced that. What's the error?